sound/soc/sof/topology.c

```diff
@@ -1441,12 +1455,15 @@ static int sof_widget_load_dai(struct snd_soc_component *scomp, int index,
 
 	/* configure dai IPC message */
 	memset(&comp_dai, 0, sizeof(comp_dai));
+
+	/* get generic component configuration from the template */
+	memcpy(&comp_dai.comp, comp, sizeof(*comp));
```

Comment:
What's wrong with:
```C
struct sof_ipc_comp_dai comp_dai = { 0 };
...
comp_dai.comp = *comp;
```
No memset, no memcpy, no sizeof, no memory safety thrill.

Reply:
@marc-hb are there any documentation guide about this for Linux kernel? I prefer the memcpy as it will copy the full memory so there will be no packing gap surprise for you (comparing to the structure assignment), I am not sure there is memory safety thrill with memcpy/memset and they are suggested not to be used in Linux kernel, we can see hundreds of example of using memcpy even in sof driver:

linux$ grep "memcpy" sound/soc/sof/* -r | wc -l
166

Reply:
If the code reads the content of alignment gaps then there is a "much bigger problem". The freedom for the optimizer not to initialize alignment gaps is not a bug, it's a feature.

The "memory safety thrill" is simply getting any memset() or memcpy() argument wrong. Most security issues are related to memory (mis)management.

It's hard to search for struct assignment, on the other hand `{ 0 }` initialization is used a lot. What is also used everywhere and seems far more common than memset/memcpy is designated initializers, e.g:

```C
struct sof_ipc_comp_dai comp_dai comp_dai = {
   .comp = {
       .hdr = {
           .size = sizeof(comp_dai),
      ....
      }
     .id = swidget->comp_id,
   
  }
}
```
It's also more readable, safe, very easy to compare with the corresponding struct definitions and probably giving the optimizer more freedom.

Reply:
I think @keyon here is just using the current conventions we have... don't know if these we're originated from some cases with variable sized members accessed through pointers etc. Anyway. let's then be logical and change it all over in sof driver if preferred, and in another PR perhaps?

Reply:
Yes, that's exactly my point, thank you @juimonen you know me well :)

Reply:
> Anyway. let's then be logical and change it all over in sof driver if preferred, and in another PR perhaps?

Massive search/replace changes break version control: git diff, merge, cherry-pick, blame, etc. and they are also tedious to review. Plus there's the risk that a newer, even better policy/style/feature comes up not long after, possibly learning from the previous one.

Unless some change is urgently needed, it's much smoother and I think much better to implement better practices (*if* agreed upon) _gradually_ when new code is added or modified anyway. So no extra churn and git noise. 

That's why tools like checkpatch look only at changed code.

Reply:
> What's wrong with:
> 
> ```c
> struct sof_ipc_comp_dai comp_dai = { 0 };
> ...
> comp_dai.comp = *comp;
> ```
> 
> No memset, no memcpy, no sizeof, no memory safety thrill.

Some static analysis tools don't work well or complain about such { 0 } assignments, so the preference has been to use an explicit memset(). It's not a statement of preference, just that when in doubt I prefer to make tools happy than argue about the merits of one solution v. the other.

Reply:
> Some static analysis tools don't work well or complain about such { 0 } assignments,

Interesting, thanks! So I guess they don't understand larger initializers either? `{ 0 }` is not really special.

It's very ironic and sad that some safety tools don't understand simpler and safer initializers built in the language and already used everywhere in the kernel but I understand the advantages of using them outweigh the rest.

Reply:
@marc-hb I'm glad to have one more group member, preferring assignments over `memset()` and `memcpy()`. If you also object to unjustified use of `kzalloc()` where a normal `kmalloc()` would do just as well, you'll be my hero of the day ;-)

Reply:
In my new solution, the extended data struct is appended to the end of the component struct and memcpy() wil use there, so mark it resolved here to make the review easier, @marc-hb if you still have similar concern, please help to raise it there.